File: 数据结构/并查集.cpp
class UnionFind {
private:
    int *parent;   // 标注当前元素的父节点的位置
    int size;       // 并查集中的元素个数

    //将父节点指向自身表示此为根结点，初始化层数1，所有元素独立成集合。
    void init() {
        for (int i = 0; i < size; i++) {
            parent[i] = i;
        }
    }
    
public:
    UnionFind (int _size): size(_size) {
        parent = new int[size];
        init();
    }
    //寻找父节点
    int find(int target) {
        if(target >= size)//数组越界
            return -1;
        if(target == parent[target])
            return target;
        parent[target] = find(parent[target])
        return parent[target];
    }
    //合并集合
    void get_together(int p, int q) {
        int pRoot = find(p);
        int qRoot = find(q);
        if(pRoot == qRoot)//本来就是同一个集合
            return;
        parent[qRoot] = pRoot;
    }
    // 如果两个节点的根节点一致则表示这两个节点是相连接的
    bool isConnected(int p, int q) {
        return find(p) == find(q);
    }
};

File: 数据结构/并查集/并查集.go
type UFSet struct {
	Parents []int
    Size []int
	N int
}

func NewUFSet(n int) UFSet {
	parents := make([]int,n)
    size := make([]int, n)
	for i:=0; i<n; i++ {
		parents[i]=i
        size[i]=1
	}
	return UFSet{
		Parents: parents,
        Size: size,
		N: n,
	}
}

func (ufs UFSet)Find(target int) int {
	if target>ufs.N {
		return -1	// 越界
	}
	if target==ufs.Parents[target] {
		return target
	}
    ufs.Parents[target] = ufs.Find(ufs.Parents[target])
	return ufs.Parents[target]
}

func (ufs UFSet)Union(p,q int) {
	pRoot,qRoot := ufs.Find(p),ufs.Find(q)
	if pRoot!=qRoot {
		ufs.Parents[pRoot] = qRoot
        ufs.Size[qRoot] += ufs.Size[pRoot]
	}
}

func main() {
	ufs := NewUFSet(10)
	ufs.Union(0,4)
	fmt.Println(ufs.Find(4),ufs.Size[ufs.Find(4)])
}